rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java

```diff
@@ -206,23 +227,32 @@ private boolean allowByMinerNotBanned(Peer sender, Message message) {
      * record event if message is repeated
      */
     private boolean allowByMessageUniqueness(Peer sender, Message message) {
-        Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
+        if (message.getMessageType() != MessageType.BLOCK_MESSAGE && message.getMessageType() != MessageType.TRANSACTIONS) {
+            return true;
+        }
 
-        boolean contains = receivedMessages.contains(encodedMessage);
+        long currentTime = System.currentTimeMillis();
 
-        if (!contains) {
-            if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
-                if (this.receivedMessages.size() >= MAX_NUMBER_OF_MESSAGES_CACHED) {
-                    this.receivedMessages.clear();
-                }
-                this.receivedMessages.add(encodedMessage);
-            }
+        synchronized (receivedPeerMessages) {
+            logger.trace("Cleaning {} messages from rlp queue", receivedPeerMessages.size());
+            receivedPeerMessages.entrySet()
+                    .stream()
+                    .filter(entry -> currentTime - entry.getValue() > RECEIVED_MESSAGES_CACHE_DURATION)
+                    .forEach(entry -> receivedPeerMessages.remove(entry.getKey()));
```

Comment:
If I'm not mistaken this will cause a `ConcurrentModificationException` as we are removing an item during the iteration.

You may want to try with `receivedPeerMessages.entrySet().removeIf`.

Reply:
Addressed